book-content/chapters/10-deriving-types.md

```diff
@@ -0,0 +1,1460 @@
+
+```ts twoslash
+const fetchUser = async (id: string) => {
+  return {
+    id,
+    name: "John Doe",
+    email: "example@email.com",
+  };
+};
+
+// ---cut---
+type User = ReturnType<typeof fetchUser>;
+//   ^?
+```
+
+In order to unwrap the `Promise` type and provide the type of the resolved value, we can use the `Awaited` utility type:
+
+```typescript
+type User = Awaited<ReturnType<typeof fetchUser>>;
+```
+
+Like before, the `User` type is now a match for the expected type, which means our test will pass as expected.
+
+It would also be possible to create intermediate types, but combining operators and type derivation gives us a more succinct solution.
+
+## Transforming Derived Types
+
+In the previous section we looked at how to derive types from functions you don't control. Sometimes, you'll also need to do the same with _types_ you don't control.
+
+### `Exclude`
+
+The `Exclude` utility type is used to remove types from a union. Let's imagine that we have a union of different states our album can be in:
+
+```typescript
+type AlbumState =
+  | {
+      type: "released";
+      releaseDate: string;
+    }
+  | {
+      type: "recording";
+      studio: string;
+    }
+  | {
+      type: "mixing";
+      engineer: string;
+    };
+```
+
+We want to create a type that represents the states that are not "released". We can use the `Exclude` utility type to achieve this:
+
+```ts twoslash
+type AlbumState =
+  | {
+      type: "released";
+      releaseDate: string;
+    }
+  | {
+      type: "recording";
+      studio: string;
+    }
+  | {
+      type: "mixing";
+      engineer: string;
+    };
+
+// ---cut---
+type UnreleasedState = Exclude<AlbumState, { type: "released" }>;
+//   ^?
+```
+
+In this case, the `UnreleasedState` type is a union of the `recording` and `mixing` states, which are the states that are not "released". `Exclude` filters out any member of the union with a `type` of `released`.
+
+We could have done it by checking for a `releaseDate` property instead:
+
+```typescript
+type UnreleasedState = Exclude<AlbumState, { releaseDate: string }>;
+```
+
+This is because `Exclude` works by pattern matching. It will remove any type from the union that matches the pattern you provide.
+
+This means we can use it to remove all strings from a union:
+
+```ts twoslash
+type Example = "a" | "b" | 1 | 2;
+
+type Numbers = Exclude<Example, string>;
+//   ^?
+```
+
+### `NonNullable`
+
+`NonNullable` is used to remove `null` and `undefined` from a type. This can be useful when extracting a type from a partial object:
+
+```ts twoslash
+type Album = {
+  artist?: {
+    name: string;
+  };
+};
+
+type Artist = NonNullable<Album["artist"]>;
+//   ^?
+```
+
+This operates similarly to `Exclude`:
+
+```typescript
+type Artist = Exclude<Album["artist"], null | undefined>;
+```
+
+But `NonNullable` is more explicit and easier to read.
```

Comment:
It might be good to call out that `NonNullable` today is defined as `T & {}` and that's because `{}` means exactly that: a non nullable value, something on which you can access properties. `'foo'.whatever` while weird and a type error on its own... it's a not a runtime error, you simply get `undefined` back from this. The same can't be said about `(null).whatever`